access.py

```diff
@@ -38,19 +47,209 @@ def __getattr__(self, item):
         raise AttributeError()
 
 
+class Output(object):
+    # GPIOs and I2C shift-register outputs
+
+    def __init__(self, address, unlock_value, open_delay):
+	if type(address) is int:
+	    GPIO.setup(address, GPIO.OUT)
+	self.address = address
+	self.unlock_value = unlock_value
+	self.open_delay = open_delay
+	self.deactivate()
+
+    def set_out(self, go_active=False):
+	new_value = go_active ^ self.unlock_value ^ 1
+	if type(self.address) is int:
+	    GPIO.output(self.address, new_value)
+	elif type(self.address) is unicode:
+	    (addr, bit) = [int(x, 16) for x in self.address.split('.')]
+	    if not config.i2c.get(addr):
+		config.i2c[addr] = 0
+	    if new_value:
+		config.i2c[addr] |= 1 << bit
+	    else:
+		config.i2c[addr] &= 1 << bit ^ 0xFF
+	    bus.write_byte_data(addr, 0x44, config.i2c[addr])	# TPIC2810 compatible
+	else:
+	    debug("gpio is unknown type: %s" % type(self.address))
+
+    def deactivate(self):
+	self.set_out(False)
+
+    def activate(self):
+	self.set_out(True)
+
+    def timed_activation(self):
+	self.activate()
+	time.sleep(self.open_delay)
+	self.deactivate()
+
+
 class CardReader(object):
-    # when scan happens, check for authorization, and fire events if successful
-    pass
+    # when scan happens, validate, check for authorization, and fire events if successful
+
+    def __init__(self, reader_config):
+        if reader_config["name"] == "<door_name>":
+            return None
+        if reader_config["is_locker"]:
+	    self.is_locker = True
+            self.lockers = []
+            for locker in reader_config["doors"]:
+                self.lockers.append(Locker(locker, self))
+        else:
+	    self.is_locker = False
+            self.door = Door(reader_config["doors"][0], self)
+	if reader_config.get("led"):
+            self.led = Output(reader_config["led"], 1, 1)
+        self.name = reader_config["name"]
+        self.stream = ""
+        self.timer = None
+        self.unlocked = False
+        self.data0 = reader_config["data0"]
+        self.data1 = reader_config["data1"]
+	self.permissions = reader_config["permissions"]
+	self.permissions.append(self.name)
+        GPIO.setup(self.data0, GPIO.IN)
+        GPIO.setup(self.data1, GPIO.IN)
+        GPIO.add_event_detect(self.data0, GPIO.FALLING,
+                              callback=self.data_pulse)
+        GPIO.add_event_detect(self.data1, GPIO.FALLING,
+                              callback=self.data_pulse)
+
+    def find_locker(self, name):
+        for locker in self.lockers:
+            if locker.name == name:
+                return locker
+        return None
```

Comment:
This golfs well: `return next((l for l in self.lockers if l.name == name), None)`.